router/qrouter/proxy_routing.go

```diff
@@ -945,110 +945,76 @@ func (qr *ProxyQrouter) routeWithRules(ctx context.Context, stmt lyx.Node, sph s
 			return nil, err
 		}
 	case *lyx.Select:
-
-		/*
-		* Ugly hack here. Code should be refactored & simplyfied.
-		* We do not error-out immidiately to check some metadata-only rouing cases.
-		 */
-		var err error
-
-		/* We cannot route SQL stmt with no FROM clause provided, but there is still
-		* a few cases to consider
-		 */
+		// We cannot route SQL statements without a FROM clause. However, there are a few cases to consider.
 		if len(node.FromClause) == 0 && (node.LArg == nil || node.RArg == nil) {
-			/* Step 1.4.8: select a_expr is routable to any shard in case when a_expr is some type of
-			data-independent expr */
-
-			spqrlog.Zero.Debug().Msg("checking any routable")
-
-			any_routable := false
 			for _, expr := range node.TargetList {
 				switch e := expr.(type) {
+				// Special cases for SELECT current_schema(), SELECT set_config(...), and SELECT pg_is_in_recovery()
 				case *lyx.FuncApplication:
-					/* case 1.4.8.1 - SELECT current_schema() special case */
-					if e.Name == "current_schema" {
-						any_routable = true
-						/* case 1.4.8.2 - SELECT set_config(...) special case */
-					} else if e.Name == "set_config" {
-						any_routable = true
-						/* case 1.4.8.3 - SELECT pg_is_in_recovery special case */
-					} else if e.Name == "pg_is_in_recovery" {
-						any_routable = true
+					if e.Name == "current_schema" || e.Name == "set_config" || e.Name == "pg_is_in_recovery" {
+						return routingstate.RandomMatchState{}, nil
 					}
+				// Expression like SELECT 1, SELECT 'a', SELECT 1.0, SELECT true, SELECT false
 				case *lyx.AExprIConst, *lyx.AExprSConst, *lyx.AExprNConst, *lyx.AExprBConst:
-					// ok
-					any_routable = true
+					return routingstate.RandomMatchState{}, nil
 
+				// Special case for SELECT current_schema
 				case *lyx.ColumnRef:
-					/* Step 1.4.8.4 - SELECT current_schema special case */
 					if e.ColName == "current_schema" {
-						any_routable = true
+						return routingstate.RandomMatchState{}, nil
 					}
-				default:
 				}
 			}
-			if any_routable {
-				return routingstate.RandomMatchState{}, nil
+		} else {
+			// Deparse populates FromClause info, so do recurse into both branches even if error encountered
+			if node.LArg != nil {
+				if err := qr.deparseShardingMapping(ctx, node.LArg, meta); err != nil {
+					return nil, err
+				}
 			}
-		} else if node.LArg != nil && node.RArg != nil {
-			/* deparse populates FromClause info,
-			* so do recurse into both branches even if error encountered
-			 */
-			l_err := qr.deparseShardingMapping(ctx, node.LArg, meta)
```

Comment:
The real reson why you needed l_err and r_err is that node.LArg and node.RArg may be nil. So, let's check them before deparseShardingMapping call.